### PR TITLE
Add JUnit xml output format

### LINKED ISSUE #963: Add support for junit xml format
It would be helpful for Continuous Integration.

### PR DESCRIPTION
**Pull request by messense**
Closes #963

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -37,6 +37,7 @@ notify = { version = "4.0.17" }
 num-bigint = { version = "0.4.3" }
 once_cell = { version = "1.16.0" }
 path-absolutize = { version = "3.0.14", features = ["once_cell_cache", "use_unix_paths_on_wasm"] }
+quick-junit = "0.3.2"
```
Comment (messense) [inline Cargo.toml:40]: https://github.com/nextest-rs/nextest/tree/main/quick-junit

**Comment by messense** [comment]
I'm not sure how to organize the output, needs more research, currently:

```bash
cargo run resources/test/fixtures/A001.py --format junit
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/ruff resources/test/fixtures/A001.py --format junit`
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="ruff" tests="4" failures="4" errors="0">
    <testsuite name="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" tests="4" disabled="0" errors="0" failures="4">
        <testcase name="Pyflakes" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="1">
            <failure message="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:1:8: `some` imported but unused" type="F401"/>
        </testcase>
        <testcase name="Pyflakes" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="2">
            <failure message="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:2:18: `some.other` imported but unused" type="F401"/>
        </testcase>
        <testcase name="Pyflakes" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="5">
            <failure message="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:5:12: Undefined name `annotation`" type="F821"/>
        </testcase>
        <testcase name="Pyflakes" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="18">
            <failure message="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:18:1: Local variable `ValueError` is assigned to but never used" type="F841"/>
        </testcase>
    </testsuite>
</testsuites>
```

**Comment by messense** [comment]
References:

* [flake8-formatter-junit-xml](https://github.com/astj/flake8-formatter-junit-xml): https://github.com/astj/flake8-formatter-junit-xml/blob/master/tests/expected.xml

**Comment by charliermarsh** [comment]
This is great! I have to look into the conventions here too.

**Comment by charliermarsh** [comment]
Another reference: https://eslint.org/docs/latest/user-guide/formatters/#junit

**Comment by charliermarsh** [comment]
So for ESLint:

- `testsuites` (name is empty)
- `testsuite package="org.ruff" name="/path/to/file.py"`
- `testcase name="org.ruff.F841" classname="/path/to/file"` (maybe `path.to.module`?)
- They don't seem to use any failure type, just the message

File: src/commands.rs [hunk 2]
```diff
@@ -61,6 +61,9 @@ pub fn explain(code: &CheckCode, format: SerializationFormat) -> Result<()> {
                 })?
             );
         }
+        SerializationFormat::Junit => {
+            todo!()
```
Comment (messense) [inline src/commands.rs:65]: I don't think it's useful to have `--explain` to output junit xml.

Comment (charliermarsh) [reply src/commands.rs:65]: I agree, log some feedback?

**Comment by messense** [comment]
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="ruff" tests="4" failures="4" errors="0">
    <testsuite name="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" tests="4" disabled="0" errors="0" failures="4">
        <testcase name="org.ruff.F401" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="1" column="8">
            <failure message="`some` imported but unused"/>
        </testcase>
        <testcase name="org.ruff.F401" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="2" column="18">
            <failure message="`some.other` imported but unused"/>
        </testcase>
        <testcase name="org.ruff.F821" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="5" column="12">
            <failure message="Undefined name `annotation`"/>
        </testcase>
        <testcase name="org.ruff.F841" file="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" line="18" column="1">
            <failure message="Local variable `ValueError` is assigned to but never used"/>
        </testcase>
    </testsuite>
</testsuites>
```

* I guess `testsuites.name` is irrelevant, but `quick-junit` requires a name so give it `ruff`.
* Should we add `end_location` to `testcase`? `location` is already added as `line` and `column`.

File: src/printer.rs [hunk 3]
```diff
@@ -96,6 +96,41 @@ impl<'a> Printer<'a> {
...
+                    let mut test_suite = TestSuite::new(filename);
+                    test_suite
+                        .extra
+                        .insert("package".to_string(), "org.ruff".to_string());
+                    for message in messages {
+                        let mut status = TestCaseStatus::non_success(NonSuccessKind::Failure);
+                        status.set_message(message.kind.body());
+                        let mut case =
+                            TestCase::new(format!("org.ruff.{}", message.kind.code()), status);
+                        case.set_classname(filename);
```
Comment (charliermarsh) [inline src/printer.rs:122]: The ESLint version strips the `.js` suffix, so maybe we should do that? I'm not sure.

**Comment by charliermarsh** [comment]
I think it's fine to omit the end location.

File: src/printer.rs (hunk 3 above)
Comment (messense) [reply src/printer.rs]:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="ruff" tests="4" failures="4" errors="0">
    <testsuite name="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" tests="4" disabled="0" errors="0" failures="4" package="org.ruff">
        <testcase name="org.ruff.F401" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="1" column="8">
            <failure message="`some` imported but unused">/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:1:8: `some` imported but unused</failure>
        </testcase>
        <testcase name="org.ruff.F401" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="2" column="18">
            <failure message="`some.other` imported but unused">/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:2:18: `some.other` imported but unused</failure>
        </testcase>
        <testcase name="org.ruff.F821" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="5" column="12">
            <failure message="Undefined name `annotation`">/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:5:12: Undefined name `annotation`</failure>
        </testcase>
        <testcase name="org.ruff.F841" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="18" column="1">
            <failure message="Local variable `ValueError` is assigned to but never used">/Users/messense/Projects/ruff/resources/test/fixtures/A001.py:18:1: Local variable `ValueError` is assigned to but never used</failure>
        </testcase>
    </testsuite>
</testsuites>
```

File: src/printer.rs [hunk 4]
```diff
@@ -96,6 +96,51 @@ impl<'a> Printer<'a> {
...
+                    let mut test_suite = TestSuite::new(filename);
+                    test_suite
+                        .extra
+                        .insert("package".to_string(), "org.ruff".to_string());
+                    for message in messages {
+                        let mut status = TestCaseStatus::non_success(NonSuccessKind::Failure);
+                        status.set_message(message.kind.body());
+                        status.set_description(format!(
+                            "{}:{}:{}: {}",
+                            filename,
```
Comment (charliermarsh) [inline src/printer.rs:122]: Can we omit the filename from the message, like in ESLint?

Comment (messense) [reply src/printer.rs]: Done.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="ruff" tests="4" failures="4" errors="0">
    <testsuite name="/Users/messense/Projects/ruff/resources/test/fixtures/A001.py" tests="4" disabled="0" errors="0" failures="4" package="org.ruff">
        <testcase name="org.ruff.F401" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="1" column="8">
            <failure message="`some` imported but unused">line 1, col 8, `some` imported but unused</failure>
        </testcase>
        <testcase name="org.ruff.F401" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="2" column="18">
            <failure message="`some.other` imported but unused">line 2, col 18, `some.other` imported but unused</failure>
        </testcase>
        <testcase name="org.ruff.F821" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="5" column="12">
            <failure message="Undefined name `annotation`">line 5, col 12, Undefined name `annotation`</failure>
        </testcase>
        <testcase name="org.ruff.F841" classname="/Users/messense/Projects/ruff/resources/test/fixtures/A001" line="18" column="1">
            <failure message="Local variable `ValueError` is assigned to but never used">line 18, col 1, Local variable `ValueError` is assigned to but never used</failure>
        </testcase>
    </testsuite>
</testsuites>
```

**Comment by charliermarsh** [comment]
Stellar, thank you!

**Pull request merged**